Make project more NixOS friendly

matt1432 opened the pull request:
This PR is mainly to allow packaging Kapowarr on NixOS for my personal use, but also makes this project use standardized python packaging that will also make it easier for other platforms to use.

- Fix reference to `beautifulsoup4` in `requirements.txt`
- Add project metadata in `pyproject.toml`
- Move all python source files to `src` directory to follow python project conventions
- Make `Kapowarr.py` a python module
- Add build configuration in `pyproject.toml`
- Use the same argv as the initial process for the subprocess since setuptools packages the script as an executable
- Get version info from `pyproject.toml` in python code instead of hardcoding it
- Update `Dockerfile` to properly install Kapowarr
- Remove version from `docker-compose.yml` since it is deprecated
- Add command line arguments for force setting log folder, download folder, port and url base
- Grey out settings in GUI that were set with command line arguments
- Don't update `pwa_manifest.json` because it always gets updated by the app at runtime anyway and also editing files in package data is bad practice

Casvt commented:
I'm sorry but I have some things to say that I cannot word without a heavy dose of Dutch bluntness, honesty and unintentional rudeness. Forgive me in advance; it's not directed to you as a person or your capabilities as a developer.

--------------

Why does no one that contributes read the [CONTRIBUTING.md](https://github.com/Casvt/Kapowarr/blob/development/CONTRIBUTING.md) file? It exists for a reason. And this PR is exactly why. It states to make a contributing request before starting on your work, with the following reasons (quote from file):

> 1. Make a [contributing request](https://github.com/Casvt/Kapowarr/issues/new?template=contribute_request.md), where you describe what you plan on doing. _This request needs to get approved before you can start._ The contributing request has multiple uses:
>     1. Avoid multiple people working on the same thing.
>     2. Avoid you wasting your time on changes that we do not wish for.
>     3. If needed, have discussions about how something will be implemented.
>     4. A place for contact, be it questions, status updates or something else.

PR's like these is why reason 2 is there. Because you did a whole lot: some welcome, most not.

1. All of the changes for the building of Kapowarr into a package (as if it's a module) are not happening. Installing Kapowarr is as simple as downloading Python if you don't have it already, installing the dependencies with `python3 -m pip install -r requirements.txt` (a very well known, standard way of doing it) and running it with `python3 Kapowarr.py`. Otherwise you can use the Docker container. And an executable with everything in it is planned. I don't see a reason for all this extra code just to be able to publish it to PyPI, which I do not have the intention of.
2. The `bs4` to `beautifulsoup4` change in the requirements is correct and welcome.
3. Simply running the subprocess with argv _seeemsssss_ correct and fine (and better than the current solution if it doesn't break anything or introduce downsides), but I'd have to check it out more to ensure it doesn't introduce bugs. Welcome.
4. Setting settings via argument parsing that you can just edit in the UI is not happening. I don't see how it really solves a problem. If you want to run with certain hosting settings, then start it up, change them and save. Kapowarr will automatically restart and apply them.
5. A setting for the log folder, that defaults to the application folder (as it is now) is a good idea and welcome.
6. _Not_ editing the manifest file is wrong, no? With your changes, the function `update_manifest` doesn't do anything. When the base URL is changed, the manifest file has to reflect that in order to fetch the image and the home page properly. You say "it always gets updated by the app at runtime anyway", and that is exactly what the `update_manifest` function is doing. When changing the base URL, it updates the manifest file too, right away when saving, so that when Kapowarr has restarted the PWA immediately works.

So there are two possible ways to continue now:
1. You revert a lot of your work such that what's left are the welcome changes, and I merge.
2. I thank you for raising the bug, various code improvements and feature ideas, and I'll implement them quickly myself.

matt1432 commented:
> Why does no one that contributes read the [CONTRIBUTING.md](https://github.com/Casvt/Kapowarr/blob/development/CONTRIBUTING.md) file? It exists for a reason.

That is a good question, through all the work I did, I never noticed it. From what I can tell, you need the file to be named `CONTRIBUTING` without the extension for it to show up before someone makes a PR: https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors

However, I don't really mind reverting most changes I did, since I learned a lot anyway.

## 1. **All of the changes for the building of Kapowarr into a package (as if it's a module) are not happening.**

Coming from NixOS, it is very odd to me to want to package a complicated program like yours as a script instead of modules.

>  just to be able to publish it to PyPI

That's not why I wanted to package it, it's to be able to make a wheel of this project and make the distribution easier on NixOS and other platforms, while also putting all the metadata in the same neat place. This also allows getting the version of the package within the code for the `about` info and reduces the amount of files to download to only the required ones for using the program.

I could probably figure out how to run it as a script on NixOS now that I'm more familiar with the code base, but I still wanted to give my thoughts on the subject.

## 3. **Simply running the subprocess with argv seeemsssss correct and fine**

I've been running my fork for a little bit and I've had no issues. However, if it is ran as a script I can't guarantee anything.

## 4. **Setting settings via argument parsing that you can just edit in the UI is not happening**

> If you want to run with certain hosting settings, then start it up, change them and save

For me this kinda caused problems because I already had an application running on the same port, so I had to stop that app and run this one to change stuff, then restart the other app. To me it just makes life easier and I don't see the downside to doing it.

## 6. **Not editing the manifest file is wrong, no?**

To be completely honest, I don't understand what that file does. The problem I had was that on NixOS, files inside packages are read-only, therefore the program stopped working when trying to edit it. Making this change made it work.

If what I did is incorrect, I'd love to learn more and find a proper solution.

Casvt commented:
4. I get your example, and I do think that we should offer a solution to this problem if possible. However, the current implementation is not the right one in my opinion. I'm not liking the whole thing of locking the settings. What I think is a better solution is the following:
    1. Using the command line argument, the user can supply a port.
    2. On startup, if a port is given, set it as the value in the settings and save it. If the value is invalid, setting it in the settings will give an error. We should catch this and report that back nicely to the user, just like we do with the database folder option.
    3. Then continue with the startup, which means that Kapowarr will start hosting on the given port (as now it's simply the value of the setting).
    4. From then on, the user use the settings like normal. The actual port used is properly displayed in the settings. If the user wants to edit it, they can like normal. However, the port value from the command line should only be applied the first time Kapowarr starts up. Otherwise port changes will not be applied as when restarting the command line port is applied again. Currently, in Kapowarr.py, there is no way to detect whether a startup is the first startup or not, so we need to add something for that.
    5. Summary: Remove the whole locking system. We simply set the port setting to the given value on first startup, and then continue like normal. We can implement the same thing for the "host" and "url base" setting.

6. The pwa_manifest.json file is for PWA support. That's the ability to add a website to your homescreen on your phone, as if it was an app. Your browser uses this file to fetch the URL of the icon of the "app" and to know what URL to actually load when starting the "app". Problem is that these URLs change when we change the base URL that Kapowarr runs on. So in order to keep these PWAs working, we need to edit the file to reflect the change in URLs.

-----------------

Also a few other points:

1. In the settings file, you're importing `environ` but it doesn't seem like you're using it. So you can probably just remove the import.
2. You added a `data_folder` setting but it's not being used. So you can just remove it.
3. The download folder should not be a command line option. Remove the option please.
4. Instead of doing `a.__setitem__(b, c)`, you can just do `a[b] = c`.

matt1432 commented:
Instead of reverting things and complicating everything, I just moved the original branch to `build-system-bak` and reset this one completely.

I've added the changes that you said were welcome:

- bs4 to beautifulsoup4 change in the requirements
- running the subprocess with argv
- setting for the log folder, that defaults to the application folder

I also changed the file name of the CONTRIBUTING file to make it popup whenever someone starts a PR.

## Features I want to implement:

### 1. First launch settings set with arguments

I really like your ideas for this.

> Currently, in Kapowarr.py, there is no way to detect whether a startup is the first startup or not, so we need to add something for that.

I'm not completely familiar with the database code, but I assume we could simply detect whether the database was created or not during the launch to know if it was the first time or not?

### 2. PWA manifest

> The pwa_manifest.json file is for PWA support. That's the ability to add a website to your homescreen on your phone, as if it was an app.

Ahh right I forgot this was a thing.

To make it work with nix, is there a way we could copy the default file on first launch (with the code that will be implemented from the previous feature) and then edit that one inside the database folder for example?

### 3. Project metadata in pyproject.toml

I know you don't want to package your project as a module, but I still think it would be nicer to have the metadata inside the `pyproject.toml` file, such as the name, version, description, etc.

We could then get the version [here](https://github.com/Casvt/Kapowarr/blob/0faca746be1abc610d64d2a7ca6d7505a465772a/backend/internals/settings.py#L77) from the `pyproject.toml` file instead of hardcoding it. It would also make it easier for thirdparty package maintainers to parse the version of your project (ie. me when I get this on NixOS).

Casvt commented:
All changes you've made up to now are approved. Bullet 3 from your most recent message (project metadata in pyproject.toml) is approved.

# 1. First launch settings set with arguments

You seem to have slightly misunderstood what I meant. I meant any first startup of Kapowarr, compared to a restart (i.e. a re-startup) initiated from within Kapowarr itself.

When supplying `--port=8080`, Kapowarr will apply it in the settings (and because of that, later in the startup, it will run on 8080 as it fetches that value from the settings). If the user now starts using Kapowarr and decides to change the port to 443 in the settings, we run into a problem. The new port value is applied in the settings, and because it's a change to the hosting settings, Kapowarr will automatically initiate a restart, triggered internally. We're now starting Kapowarr up on port 443, as we set in the settings. But oh wait, we still have the command line option saying it should be 8080 from the very first time we ourselves actually manually started it up, so it will set the port to 8080 in the settings and startup on 8080 again instead.

So we need to detect whether it's the first startup (where the user actually starts Kapowarr), because only then do we process/apply the port command line option. Once Kapowarr restarts internally, the port command line option should be ignored.

# 2. PWA manifest

The file has to be edited every time a change to the hosting settings is made, like it is now. The problem you're having is that you're running in an environment where files can not be written to. Moving the file doesn't seem like it would solve anything, as after editing it you'd have to move it back to the original folder it was in and that sounds to me like a write (if editing a file is a write, then adding an entire file is surely a write too). And it has to be in the `frontend/static` folder because that is the only folder that is allowed to serve public static web-UI files. I'll see if I can fix the whole editing of the manifest file completely, but there's probably no way around it. So the only feasible solution might be to just figure out how to allow writing to the files in your environment.

matt1432 commented:
Turns out running the subprocess with argv only works when the project is packaged as a module, so I dropped that change.

For the metadata, I still need to implement getting the version in the python code.

## 1. First launch settings set with arguments

Ahh I see, I like this.

## 2. PWA manifest

Could we add an argument to specify the static files folder? Then I could just have the NixOS service copy the files from the package and edit them in the specified location.

Casvt commented on backend/base/logging.py:
```diff
@@ -100,10 +101,18 @@ def shouldRollover(self, record: logging.LogRecord) -> int:
-    LOGGING_CONFIG["handlers"]["file"]["filename"] = get_log_filepath()
+    if log_folder is None:
+        from backend.base.files import folder_path
+        LOGGING_CONFIG["handlers"]["file"]["filename"] = folder_path(Constants.LOGGER_FILENAME)
```
This can be moved to the top of the file

Casvt commented on backend/base/logging.py:
```diff
@@ -100,10 +101,18 @@ def shouldRollover(self, record: logging.LogRecord) -> int:
 
-    LOGGING_CONFIG["handlers"]["file"]["filename"] = get_log_filepath()
+    if log_folder is None:
+        from backend.base.files import folder_path
```
This can be moved to the top of the file

matt1432 replied:
When I do, it gives me this error
```console
Apr 20 16:27:51 nos kapowarr[3855938]:     from backend.base.helpers import check_python_version, get_python_exe
Apr 20 16:27:51 nos kapowarr[3855938]:   File "/nix/store/bqx9an1dzksqinlavh1f4hd3063i7q4i-kapowarr-1.1.1+dirty/lib/python3.12/site-packages/backend/base/helpers.py", line 28, in <module>
Apr 20 16:27:51 nos kapowarr[3855938]:     from backend.base.logging import LOGGER, get_log_filepath
Apr 20 16:27:51 nos kapowarr[3855938]:   File "/nix/store/bqx9an1dzksqinlavh1f4hd3063i7q4i-kapowarr-1.1.1+dirty/lib/python3.12/site-packages/backend/base/logging.py", line 10, in <module>
Apr 20 16:27:51 nos kapowarr[3855938]:     from backend.base.files import folder_path
Apr 20 16:27:51 nos kapowarr[3855938]:   File "/nix/store/bqx9an1dzksqinlavh1f4hd3063i7q4i-kapowarr-1.1.1+dirty/lib/python3.12/site-packages/backend/base/files.py", line 16, in <module>
Apr 20 16:27:51 nos kapowarr[3855938]:     from backend.base.helpers import check_filter, force_suffix
Apr 20 16:27:51 nos kapowarr[3855938]: ImportError: cannot import name 'check_filter' from partially initialized module 'backend.base.helpers' (most likely due to a circular import) (/nix/store/bqx9an1dzksqinlavh1f4hd3063i7q4i-kapowarr-1.1.1+dirty/lib/python3.12/site-packages/backend/base/helpers.py)
```

Casvt replied:
Oh yeah that's right sorry

Casvt commented on backend/base/logging.py:
```diff
@@ -100,10 +101,18 @@ def shouldRollover(self, record: logging.LogRecord) -> int:
+        from backend.base.files import folder_path
+        LOGGING_CONFIG["handlers"]["file"]["filename"] = folder_path(Constants.LOGGER_FILENAME)
+    else:
+        LOGGING_CONFIG["handlers"]["file"]["filename"] = join(log_folder, Constants.LOGGER_FILENAME)
```
No validation is done whether the log folder exists, and whether it's a folder (and not a file). If that's not the case, it should raise a Value error that's caught in Kapowarr.py and processed at the same place where the database folder error is caught.

Casvt replied:
Mate please read my entire message. You forgot half of it:

> [...] that's caught in Kapowarr.py and processed at the same place where the database folder error is caught.

Casvt commented:
Getting the version from the metadata file into the about_data should be like two lines of code so I'll assume that's done.

1. I'm currently working on another project where I already implemented this, but I haven't pushed it yet. I'll share my implementation in a few hours when I'm at my pc.
2. I found a method to avoid having to edit the file at all. I'll implement it later myself, so that point we can cross off.

I also left some comments on specific lines.

Casvt commented:
Please at this point just stop. Sorry but my patience has run out. I've probably spent three times as much time guiding you as I would have implementing this stuff myself.

You casually changed the minimum required Python version for the project from 3.8 to 3.11. That's a big no-no, especially in this case. Upgrading the python version is a braking change, which would require me to release a new major version (V2.x.x) and notify all users in every way possible and well in advance that they'll have to manually upgrade Kapowarr. And all that because instead of 7 lines of code you want to use 1. The following works just fine as well (wrote this recently for my other project):

```python
with open(folder_path("pyproject.toml"), "r") as f:
    for line in f:
        if line.startswith("version = "):
            version = "v" + line.split('"')[1]
            break
    else:
        raise RuntimeError("Version not found in pyproject.toml")
```

And I already have a working implementation of the command line arguments that I can just copy and paste. And I have a solution ready for the editing of the manifest file.

So I will definitely be taking over a part of the changes, and I thank you for the suggestions and reports, but I'll be ending it here.
